src/services/TestResultsService.ts

```diff
@@ -109,52 +125,126 @@ export class TestResultsService {
     return testResults;
   }
 
-  public insertTestResult(payload: ITestResultPayload) {
-    let validationSchema = null;
-    if (Object.keys(payload).length === 0) { // if empty object
-      return Promise.reject(new HTTPError(400, ERRORS.PayloadCannotBeEmpty));
+  public updateTestResult(systemNumber: string, payload: ITestResult, msUserDetails: IMsUserDetails) {
+    this.removeNonEditableAttributes(payload);
+    let validationSchema = this.getValidationSchema(payload.vehicleType, payload.testStatus);
+    validationSchema = validationSchema!.keys({countryOfRegistration: Joi.string().valid(countryOfRegistration).required().allow("", null)});
+    validationSchema = validationSchema!.optionalKeys(["testStationType", "testerEmailAddress", "testEndTimestamp", "systemNumber", "vin"]);
+    const validation: ValidationResult<any> | any | null = Joi.validate(payload, validationSchema);
+
+    if (validation !== null && validation.error) {
+      return Promise.reject(new HTTPError(400,
+          {
+            errors: validation.error.details.map((detail: { message: string; }) => {
+              return detail.message;
+            })
+          }));
     }
+    return this.testResultsDAO.getBySystemNumber(systemNumber)
+        .then((result) => {
+          const response: ITestResultData = {Count: result.Count, Items: result.Items};
+          const testResults = this.checkTestResults(response);
+          const oldTestResult = this.getTestResultToArchive(testResults, payload.testResultId);
+          const newTestResult: ITestResult = cloneDeep(oldTestResult);
+          newTestResult.testVersion = TEST_VERSION.CURRENT;
+          oldTestResult.testVersion = TEST_VERSION.ARCHIVED;
+          mergeWith(newTestResult, payload);
```

Comment:
You're not using a customiser, so merge is more suitable than mergeWith

Reply:
Also you might want to be careful with how merge/mergeWith handles array merges. I'm not sure exactly what data you intend to send down but if you intend replacements of anything in arrays, this will not do that.

Reply:
For reference: https://lodash.com/docs/#mergeWith

Reply:
I think I'll need to use a customiser at some point because of the array changes, wasn't specifically mentioned in this ticket though, I think we'll have another ticket for removing defects and test-types and there I'll definitely need an array customiser